File: mvp/sdk/typescript/src/api/transcription.ts
import { BaseAPI } from "./base";
import {
  Transcription,
  TranscriptionCreateRequest,
  SmartTranscribeRequest,
  SmartTranscribeResponse,
  TranscriptionWithSpeakers,
  Chapter,
  Summary,
  Keyword,
  SearchTranscriptionResult,
  BatchTranscribeResponse,
  YouTubeMetadata,
  PlaylistTranscribeRequest,
  PlaylistTranscribeResponse,
  LiveStreamStatusResponse,
  LiveStreamCaptureRequest,
  LiveStreamCaptureResponse,
  VideoAnalyzeRequest,
  VideoAnalyzeResponse,
  PaginatedResponse,
  PaginationParams,
} from "../types";

/**
 * Transcription API — YouTube / audio / video transcription with speaker
 * diarization, chaptering, export, batch processing and live-stream capture.
 */
export class TranscriptionAPI extends BaseAPI {
  // -- CRUD ----------------------------------------------------------------

  /** Create a transcription job from a URL (YouTube, audio, video). */
  async create(data: TranscriptionCreateRequest): Promise<Transcription> {
    return this._post<Transcription>("/api/transcription/", data);
  }

  /** Upload an audio/video file for transcription. */
  async upload(file: Blob | File, language?: string): Promise<Transcription> {
    const form = new FormData();
    form.append("file", file);
    if (language) form.append("language", language);
    return this._request<Transcription>("/api/transcription/upload", {
      method: "POST",
      body: form,
    });
  }

  /** List transcriptions (paginated). */
  async list(params?: PaginationParams): Promise<PaginatedResponse<Transcription>> {
    return this._get<PaginatedResponse<Transcription>>("/api/transcription/", {
      limit: params?.limit,
      offset: params?.offset,
    });
  }

  /** Get transcription by ID. */
  async get(id: string): Promise<Transcription> {
    return this._get<Transcription>(`/api/transcription/${id}`);
  }

  /** Delete a transcription. */
  async delete(id: string): Promise<void> {
    return this._delete(`/api/transcription/${id}`);
  }

  /** Get transcription statistics. */
  async stats(): Promise<Record<string, unknown>> {
    return this._get<Record<string, unknown>>("/api/transcription/stats");
  }

  // -- Speaker diarization ------------------------------------------------

  /** Get transcription with speaker diarization. */
  async speakers(id: string): Promise<TranscriptionWithSpeakers> {
    return this._get<TranscriptionWithSpeakers>(`/api/transcription/${id}/speakers`);
  }

  // -- Smart transcription ------------------------------------------------

  /** Smart transcription with automatic provider routing. */
  async smartTranscribe(data: SmartTranscribeRequest): Promise<SmartTranscribeResponse> {
    return this._post<SmartTranscribeResponse>("/api/transcription/smart-transcribe", data);
  }

  // -- Chaptering & summary -----------------------------------------------

  /** Auto-generate chapters with summaries. */
  async chapters(id: string): Promise<{ chapters: Chapter[] }> {
    return this._post<{ chapters: Chapter[] }>("/api/transcription/auto-chapter", {
      transcription_id: id,
    });
  }

  /** Generate a summary of a transcription. */
  async summary(id: string, style?: string): Promise<Summary> {
    return this._post<Summary>(`/api/transcription/${id}/summary`, { style });
  }

  /** Extract keywords from a transcription. */
  async keywords(id: string): Promise<Keyword[]> {
    return this._get<Keyword[]>(`/api/transcription/${id}/keywords`);
  }

  // -- Export --------------------------------------------------------------

  /** Export transcription in the given format. */
  async export(id: string, format: "srt" | "vtt" | "txt" | "md" | "json"): Promise<string> {
    return this._get<string>(`/api/transcription/${id}/export`, { format });
  }

  // -- Batch ---------------------------------------------------------------

  /** Submit batch transcription jobs. */
  async batch(urls: string[]): Promise<BatchTranscribeResponse> {
    return this._post<BatchTranscribeResponse>("/api/transcription/batch", { urls });
  }

  // -- Metadata ------------------------------------------------------------

  /** Extract YouTube video metadata. */
  async metadata(url: string): Promise<YouTubeMetadata> {
    return this._post<YouTubeMetadata>("/api/transcription/metadata", { url });
  }

  // -- Playlist ------------------------------------------------------------

  /** Transcribe an entire YouTube playlist. */
  async playlist(data: PlaylistTranscribeRequest): Promise<PlaylistTranscribeResponse> {
    return this._post<PlaylistTranscribeResponse>("/api/transcription/playlist", data);
  }

  // -- Live stream ---------------------------------------------------------

  /** Check if a live stream is active. */
  async streamStatus(url: string): Promise<LiveStreamStatusResponse> {
    return this._post<LiveStreamStatusResponse>("/api/transcription/stream/status", { url });
  }

  /** Capture a segment of a live stream. */
  async streamCapture(data: LiveStreamCaptureRequest): Promise<LiveStreamCaptureResponse> {
    return this._post<LiveStreamCaptureResponse>("/api/transcription/stream/capture", data);
  }

  // -- Video analysis ------------------------------------------------------

  /** Analyze video frames with Vision AI. */
  async analyzeVideo(data: VideoAnalyzeRequest): Promise<VideoAnalyzeResponse> {
    return this._post<VideoAnalyzeResponse>("/api/transcription/video/analyze", data);
  }

  // -- Search --------------------------------------------------------------

  /** Search across all transcriptions. */
  async search(query: string, limit?: number): Promise<SearchTranscriptionResult[]> {
    return this._get<SearchTranscriptionResult[]>("/api/transcription/search", {
      query,
      limit,
    });
  }
}

File: mvp/sdk/python/saas_ia/api/transcription.py
"""Transcription API — YouTube / audio / video transcription with speaker
diarization, chaptering, export, batch processing, and live-stream capture."""

from __future__ import annotations

from typing import Any, BinaryIO

from saas_ia.api.base import BaseAPI


class TranscriptionAPI(BaseAPI):
    """Wraps ``/api/transcription`` endpoints."""

    # -- CRUD ---------------------------------------------------------------

    async def create(
        self,
        source_url: str,
        *,
        language: str | None = None,
    ) -> dict[str, Any]:
        """Create a transcription job from a URL."""
        body: dict[str, Any] = {"source_url": source_url}
        if language:
            body["language"] = language
        return await self._post("/api/transcription/", json=body)

    async def upload(
        self,
        file: BinaryIO,
        filename: str = "upload",
        *,
        language: str | None = None,
    ) -> dict[str, Any]:
        """Upload an audio/video file for transcription."""
        files = {"file": (filename, file)}
        data = {"language": language} if language else None
        return await self._post("/api/transcription/upload", data=data, files=files)

    async def list(
        self,
        *,
        limit: int | None = None,
        offset: int | None = None,
    ) -> dict[str, Any]:
        """List transcriptions (paginated)."""
        return await self._get(
            "/api/transcription/",
            params={"limit": limit, "offset": offset},
        )

    async def get(self, job_id: str) -> dict[str, Any]:
        """Get transcription by ID."""
        return await self._get(f"/api/transcription/{job_id}")

    async def delete(self, job_id: str) -> None:
        """Delete a transcription."""
        await self._delete(f"/api/transcription/{job_id}")

    async def stats(self) -> dict[str, Any]:
        """Get transcription statistics."""
        return await self._get("/api/transcription/stats")

    # -- Speaker diarization ------------------------------------------------

    async def speakers(self, job_id: str) -> dict[str, Any]:
        """Get transcription with speaker diarization."""
        return await self._get(f"/api/transcription/{job_id}/speakers")

    # -- Smart transcription ------------------------------------------------

    async def smart_transcribe(
        self,
        source_url: str,
        *,
        language: str | None = None,
        enable_diarization: bool = False,
    ) -> dict[str, Any]:
        """Smart transcription with automatic provider routing."""
        body: dict[str, Any] = {"source_url": source_url}
        if language:
            body["language"] = language
        if enable_diarization:
            body["enable_diarization"] = True
        return await self._post("/api/transcription/smart-transcribe", json=body)

    # -- Chaptering & summary -----------------------------------------------

    async def chapters(self, job_id: str) -> dict[str, Any]:
        """Auto-generate chapters with summaries."""
        return await self._post(
            "/api/transcription/auto-chapter",
            json={"transcription_id": job_id},
        )

    async def summary(
        self,
        job_id: str,
        *,
        style: str | None = None,
    ) -> dict[str, Any]:
        """Generate a summary of a transcription."""
        body: dict[str, Any] = {}
        if style:
            body["style"] = style
        return await self._post(f"/api/transcription/{job_id}/summary", json=body)

    async def keywords(self, job_id: str) -> list[dict[str, Any]]:
        """Extract keywords from a transcription."""
        return await self._get(f"/api/transcription/{job_id}/keywords")

    # -- Export -------------------------------------------------------------

    async def export(
        self,
        job_id: str,
        format: str = "txt",
    ) -> str:
        """Export transcription in the given format (srt, vtt, txt, md, json)."""
        return await self._get(
            f"/api/transcription/{job_id}/export",
            params={"format": format},
        )

    # -- Batch --------------------------------------------------------------

    async def batch(self, urls: list[str]) -> dict[str, Any]:
        """Submit batch transcription jobs."""
        return await self._post("/api/transcription/batch", json={"urls": urls})

    # -- Metadata -----------------------------------------------------------

    async def metadata(self, url: str) -> dict[str, Any]:
        """Extract YouTube video metadata."""
        return await self._post("/api/transcription/metadata", json={"url": url})

    # -- Playlist -----------------------------------------------------------

    async def playlist(
        self,
        playlist_url: str,
        *,
        language: str | None = None,
        max_videos: int | None = None,
    ) -> dict[str, Any]:
        """Transcribe an entire YouTube playlist."""
        body: dict[str, Any] = {"playlist_url": playlist_url}
        if language:
            body["language"] = language
        if max_videos:
            body["max_videos"] = max_videos
        return await self._post("/api/transcription/playlist", json=body)

    # -- Live stream --------------------------------------------------------

    async def stream_status(self, url: str) -> dict[str, Any]:
        """Check if a live stream is active."""
        return await self._post(
            "/api/transcription/stream/status",
            json={"url": url},
        )

    async def stream_capture(
        self,
        url: str,
        *,
        duration_seconds: int | None = None,
    ) -> dict[str, Any]:
        """Capture a segment of a live stream."""
        body: dict[str, Any] = {"url": url}
        if duration_seconds:
            body["duration_seconds"] = duration_seconds
        return await self._post("/api/transcription/stream/capture", json=body)

    # -- Video analysis -----------------------------------------------------

    async def analyze_video(
        self,
        source_url: str,
        *,
        interval_seconds: int | None = None,
        max_frames: int | None = None,
    ) -> dict[str, Any]:
        """Analyze video frames with Vision AI."""
        body: dict[str, Any] = {"source_url": source_url}
        if interval_seconds:
            body["interval_seconds"] = interval_seconds
        if max_frames:
            body["max_frames"] = max_frames
        return await self._post("/api/transcription/video/analyze", json=body)

    # -- Search -------------------------------------------------------------

    async def search(
        self,
        query: str,
        *,
        limit: int | None = None,
    ) -> list[dict[str, Any]]:
        """Search across all transcriptions."""
        return await self._get(
            "/api/transcription/search",
            params={"query": query, "limit": limit},
        )
